README.md
# bunbun

Node.JS based simple, lightweight task bunner... I mean runner

<p align="center">
  <img src="bunbun.png" />
</p>

## Why?

⏱ **Lightweight and very fast** - very simple in design, without any library compilers that you will never use, just plain task runner

📐 **Ready to use** - just add to your project, create file and prepare your first task

🔌 **Extendable** - "extends-by-code" philosophy - you are owner of your build flow, without any magic and invisible parts/configurations etc., neither no compilers included

✨ **Universal** - this library contains **only** language/tool/framework agnostic tools, all compilations/copying things are in your hands

🎈 **Without magic** - all of your tasks are just JS code, without any configs, 3rd-party plugins etc. just plain source code, therefore the learning curve is very low

🐞 **Debuggable** - debugging is very simple because none of all tasks are done invisibly underhood, it's easy to find erroring part if you see all steps of build process, also you can enable debugging mode in Bunbun to log all things

🧰 **Safe versioning** - version of Bunbun will never break any of building processes because Bunbun don't conains any API for them, updating Bunbun will need from you only update your building script - it's all, you will never have to wait for a some plugin to be updated

## Requirements

- `Node.JS >= 7.6v`
    - ...because of default support of async/await syntax
- or `Node.JS >= 10v`
    - ...for execution and handling processes
- `NPM`

## How to install?

Simplest way is just using npm:

```bash
npm install --save-exact --save-dev bunbun
```

## Real example

<details>
  <summary>📚 Click to expand the sample code</summary>

Example code:

```js
const { build } = require('esbuild');
const nodeSass = require('node-sass');
const Bunbun = require('./../dist/main.js');

const sass = args => new Promise((res, rej) => {
    nodeSass.render(args, (err, data) => {
        err ? rej(err) : res(data);
    });
});

const hash = async file => {
    const res = await $.fs.hash(file, 'md5', 'base64');
    return res.replace(/[^a-z0-9]/gi, '');
};

const SOURCE_EXTS = '(scss|ts|tsx|js|jsx|html)';

const $ = new Bunbun;

$.task('build:typescript', async () => {
    await $.until('hash');
    await build({
        sourcemap: true,
        format: 'iife',
        minify: true,
        bundle: true,
        outfile: './build/index.js',
        entryPoints: ['./src/index.tsx'],
        platform: 'browser',
        tsconfig: './tsconfig.json',
        define: {
            'process.env.NODE_ENV': '"develop"',
        },
    });
    $.run('hash');
});

$.task('build:scss', async () => {
    await $.until('hash');
    const result = await sass({
        file: './src/index.scss',
        sourceMap: './index.css.map',
        outFile: './index.css',
        outputStyle: 'compressed',
        sourceMapContents: true,
    });
    await $.fs.write('./build/index.css', result.css || '');
    await $.fs.write('./build/index.css.map', result.map || '');
    $.run('hash');
});

$.task('hash', async () => {
    await $.fs.edit('./src/index.html', async html => {
        const jsHash = await hash('./build/index.js');
        const cssHash = await hash('./build/index.css');
        return html
            .replace('__JS_HASH__', jsHash)
            .replace('__CSS_HASH__', cssHash);
    });
});

$.alias('build', ['build:typescript', 'build:scss']);

$.task('assets', async () => {
    let list = await $.fs.list(['./src/**/*.*', `!**/*.${SOURCE_EXTS}`]);
    for (const x of list) {
        await $.fs.copy(x, x.replace(/^\.\/src/, './build'));
    }
});

$.task('watch', async () => {
    const server = $.serve('./build', 8080);

    $.fs.watch(`./src/**/*.${SOURCE_EXTS}`, async () => {
        await $.run('build');
        server.reload();
    });

    $.fs.watch(`./src/**/*.!${SOURCE_EXTS}`, async () => {
        await $.run('assets');
        server.reload();
    });
});

$.start('build');
```

</details>


## Table of Contents

- [`class Bunbun`](#class-bunbun)
    - [`fs`](#class-bunbun--fs) - filesystem API
    - [`logger`](#class-bunbun--logger) - logger API
    - [`alias()`](#class-bunbun--alias) - regiser alias of multiple tasks
    - [`debounce()`](#class-bunbun--debounce) - allows to debounce by promise and time
    - [`exec()`](#class-bunbun--exec) - execute terminal command
    - [`hash()`](#class-bunbun--hash) - hash given file
    - [`rescue()`](#class-bunbun--rescue) - catch exception making result optional
    - [`run()`](#class-bunbun--run) - run registered task
    - [`start()`](#class-bunbun--start) - run tasks by `process.argv`
    - [`serve()`](#class-bunbun--serve) - create new http server
    - [`task()`](#class-bunbun--task) - register new task
    - [`until()`](#class-bunbun--until) - waits until task has been done
    - [`wait()`](#class-bunbun--wait) - `setTimeout` but in `await`/`async` convention
- [`class Logger`](#class-logger)
    - [`debugging`](#class-logger--debugging) - determine if logger have to be louder
    - [`silent`](#class-logger--silent) - determine if logger have to log nothing
    - [`debug()`](#class-logger--debug) - debug log
    - [`error()`](#class-logger--error) - error log
    - [`format()`](#class-logger--format) - prepare message before log
    - [`log()`](#class-logger--log) - basic log
    - [`success()`](#class-logger--success) - success log
- [`class Fs`](#class-fs)
    - [`cwd`](#class-fs--cwd) - current working directory
    - [`copy()`](#class-fs--copy) - copy (also recursive and glob option) files/dirs
    - [`createDir()`](#class-fs--createdir) - create dir
    - [`createTempDir()`](#class-fs--createtempdir) - create unique temporary dir
    - [`edit()`](#class-fs--edit) - reads and writes back file
    - [`exists()`](#class-fs--exists) - check whenever file/dir exists and returns type
    - [`hash()`](#class-fs--hash) - hash given file
    - [`list()`](#class-fs--list) - list files/dirs
    - [`read()`](#class-fs--read) - reads data from file
    - [`remove()`](#class-fs--remove) - removes file/dir
    - [`rename()`](#class-fs--rename) - renames (also moves) file/dir
    - [`watch()`](#class-fs--watch) - watches files/dirs for changes
    - [`write()`](#class-fs--write) - writes data into file
- [`class HttpServer`](#class-httpserver)
    - [`reload()`](#class-httpserver--reload) - reloads all html pages

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun"><code>class Bunbun</code></h3>

Main class of whole script, it's also default exported thing from module

```typescript
type Options = {
    // Determines if logger should be louder by default
    debugging: boolean; // default: false

    // Determines if logger should be silent by default
    // this option has higher priority than `debugging`
    silent: boolean; // default: false

    // Determines if tasks should wait for previous promise by default
    debouncePromise: boolean; // default: true

    // Determines if tasks should debounce and with what time by default
    // 0 means turn off
    debounceTime: number; // default: 200

    // Current working directory for fs/exec etc.
    cwd: string; // default: process.cwd()
};

type Bunbun = {
    new(options?: Partial<Options>);
    // ...
};
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```typescript
const Bunbun = require('bunbun');

// Debugging instance
const $1 = new Bunbun({
    debugging: true,
});

// Default instance
const $2 = new Bunbun();
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--fs"><code>bunbun.fs</code></h3>

[`Fs`](#class-fs) (filesystem) instance available for external usage

```typescript
fs: Fs;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.fs.read('./file.txt').then(data => {
    // ...
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--logger"><code>bunbun.logger</code></h3>

[`Logger`](#class-logger) instance used by Bunbun and available for external usage as well

```typescript
logger: Logger;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

const myVar = { test: 1 };

$.logger.log('my var = $$', myVar);
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--alias"><code>bunbun.alias()</code></h3>

Registers alias of tasks, those tasks will be executed asynchronously. Name of such alias will be in same pool where tasks are so you can't register task of same name which already exists

```typescript
alias(
    name: string,
    tasks: string[],
): void;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('foo', () => {
    console.log('foo');
});
$.task('bar', async () => {
    await someLongAction();
    console.log('bar');
});

$.alias('baz', ['bar', 'foo']);
// is equal to
$.task('qux', async () => {
    return Promise.all([
        $.run('bar'),
        $.run('foo'),
    ]);
});

$.run('baz'); // or $.run('qux');
// output:
// >> foo
// >> bar (because it's waited for long running function)
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--debounce"><code>bunbun.debounce()</code></h3>

Creates debounce function of given function, this kind of function allows you avoid multiple calls of time-critical functions, e.g:
- To avoid overloop building process for same code
- To avoid overwriting things at overwriting they in previous call

> 💡 **Tip** - time debouncing don't call function until given timeout, promise debouncing call function and wait for fulfilling last promise

```typescript
debounce<T>(
    fn: () => Promise<T> | T,
    time: number = 0,
): Promise<T>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

// Not working debounce (without promise nor time)
let i1 = 0;
const fnNothing = $.debounce(() => {
    console.log('1', i1 += 1);
});
// Promise based debounce
let i2 = 0;
const fnPromise = $.debounce(() => {
    new Promise(res => setTimeout(() => {
        console.log('2', i2 += 1);
        res();
    }, 10 * 1000))
});
// Time based debounce
let i3 = 0;
const fnTime = $.debounce(() => {
    console.log('3', i3 += 1);
}, 10 * 1000);
// Promise and time based debounce
let i4 = 0;
const fnPromiseAndTime = $.debounce(() =>
    new Promise(res => setTimeout(() => {
        console.log('4', i4 += 1);
        res();
    }, 9 * 1000))
, 9 * 1000);

// --- results ---

fnNothing(); // > 1
fnNothing(); // > 2
fnNothing(); // > 3
fnNothing(); // > 4

fnPromise(); // > wait for promise > 1
// after 5s
fnPromise(); // > wait for previous result > 1
// after 5s
fnPromise(); // > wait for new promise > 2
// after 10s
fnPromise(); // > wait for new promise > 3

fnTime(); // > wait for timeout > 1
// after 5s
fnTime(); // > wait for previous timeout > 1
// after 5s
fnTime(); // > wait for new timeout > 2
// after 10s
fnTime(); // > wait for new timeout > 3

fnPromiseAndTime(); // > wait for timeout > wait for promise > 1
// after 5s
fnPromiseAndTime(); // > wait for previous timeout > wait for previous promise > 1
// after 5s
fnPromiseAndTime(); // > wait for previous promise > 1
// after 10s
fnPromiseAndTime(); // > wait for new timeout > wait for new promise > 2
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--exec"><code>bunbun.exec()</code></h3>

Allows you to execute command in shell and wait for result

> 💡 **Tip** - if timeout is equal or less than 0 then timeout will don't be turned on

```typescript
exec(
    command: string,
    timeout: number = 0,
): Promise<{
    stdout: string;
    stderr: string;
}>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.exec('node -v').then(({ stdout }) => {
    console.log(stdout);
});

// or

$.task('node-v', async () => {
    const { stdout } = await $.exec('node -v');
    console.log(stdout);
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--hash"><code>bunbun.hash()</code></h3>

Hash given string

```typescript
hash(
    text: string,
    algorithm: 'md5' | 'sha1' | 'sha256' | 'sha512' = 'md5',
    encoding: 'hex' | 'base64' | 'buffer' | 'latin1' = 'base64'
): Promise<string>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.hash('test').then(hash => console.log(hash));
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--rescue"><code>bunbun.rescue()</code></h3>

Allows catch promise and keep `async`/`await` syntax

> 💡 **Tip** - be careful and never call (`await `)`rescue(await someFunc())` because this `someFunc()` will throw in higher, current context instead of `rescue`'s context, instead of that call `await rescue(someFunc())`

> 🎈 **Fun fact** - in Ruby language `rescue` keyword is used instead of typical `catch` keyword - that's why Bunbun use this name in this context

```typescript
rescue<T1, T2>(
    promise: Promise<T1> | T1,
    alter?: T2,
): Promise<T1 | T2>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

const justReject = async () => {
    throw 'wowie, such error';
};

const dontReject = async () => 'ok!';

$.task('test-1', async () => {
    const error = await $.rescue(justReject());
    console.log(error); // > 'wowie, such error'
});

$.task('test-2', async () => {
    const error = await $.rescue(await justReject());
    // unreachable code because `justReject` is executed
    // in current context!
});

$.task('test-3', async () => {
    const ok = await $.rescue(dontReject());
    console.log(ok); // > 'ok!'
});

$.task('test-4', async () => {
    const ok = await $.rescue('ok!');
    console.log(ok); // > 'ok!'
});

$.task('test-5', async () => {
    const alternativeValue = await $.rescue(justReject(), 'ok!');
    console.log(alternativeValue); // > 'ok!'
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--run"><code>bunbun.run()</code></h3>

Run single task by name and returns promise of call of this task

> 💡 **Tip** - if task with such name will be unable to found then this function will throw `false`

```typescript
run(name: string): Promise<boolean>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('reject', async () => {
    throw 'oh no!';
});

$.task('resolve', async () => {
    return 'ok!';
});

$.task('test', async () => {
    // try run task
    await $.rescue($.run('reject'));
    console.log(
        await $.rescue($.run('reject')
    ); // > 'oh no!'
    console.log(
        await $.rescue($.run('?')
    ); // > false
    console.log(
        await $.rescue($.run('resolve')
    ); // > true
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--start"><code>bunbun.start()</code></h3>

Similar to [`bunbun.run()`](#class-bunbun--run) but instead of running single task allows run multiple tasks, with single fallback task. This function will don't throw at error (only log it)

> 💡 **Tip** - this function is created mainly for using as main method

```typescript
start(
    defaultTask: string = 'default',
    tasks: string[] = process.argv.slice(2),
): Promise<void>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('reject', async () => {
    throw 'oh no!';
});

$.task('resolve', async () => {
    return 'ok!';
});

$.start(); // run 'default' task or given names from terminal
// or
$.start('resolve'); // run 'resolve' task or given names from terminal
// or
$.start(
    'resolve',
    ['resolve', 'reject'],
); // run always 'resolve' and 'reject'
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--serve"><code>bunbun.serve()</code></h3>

Creates new instance of [`class HttpServer`](#class-httpserver)

```typescript
type Options = {
    fallback: string = './index.html',,
    reload: boolean = true,
    reloadPort: number = 8181,
};

serve(
    directory: string,
    port: number,
    options?: Partial<Options>,
): HttpServer;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

const server = $.serve('build', 8080);

// after some action you can force reload of pages
server.reload();
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--task"><code>bunbun.task()</code></h3>

Register new task under given name. Task can be simple function or async function if needed

> 💡 **Tip** - if you cannot use `async`/`await` syntax then returning `Promise` instance will give same effect

```typescript
task(
    name: string,
    fn: () => unknown,
): void;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('foo', () => {
    console.log('foo');
});
$.task('bar', async () => {
    await someLongAction();
    console.log('bar');
});
$.task('baz', async () => {
    await $.run('foo');
    await $.run('bar');
});
$.task('qux', async () => {
    await $.run('bar');
    await $.run('foo');,
});

$.run('baz');
// output:
// >> foo
// >> bar (because it's waited for long running function)

$.run('qux');
// output:
// >> bar
// >> foo (because it's waited until previous task ended)
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--until"><code>bunbun.until()</code></h3>

Waits until given task has been completed, it doesn't matter if it ended successfully, throws if task has been not found

```typescript
until(name: string): Promise<boolean>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('long', async () => {
    await $.wait(1500);
});
$.task('run-and-wait', async () => {
    $.run('long');
    await $.until('long');
    // waits until 'long' ended
    await $.until('long');
    // pass immediately because 'long' is not running
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-bunbun--wait"><code>bunbun.wait()</code></h3>

`setTimeout()` alternative for `async`/`await` syntax

```typescript
wait(time: number): Promise<void>;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```javascript
const $ = new Bunbun;

$.task('long', async () => {
    console.log('wait moment');
    await $.wait(1500);
    console.log('hello again');
});
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-logger"><code>class Logger</code></h3>

Class prepared for logging purposes. Logger is already constructed with [`class Bunbun`](#class-bunbun) with default settings

```typescript
type Logger = {
    new();
    // ...
};
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```typescript
const $ = new Bunbun();
$.logger; // here you are
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-logger--debugging"><code>logger.debugging</code></h3>

Determines if logger should log more useful data for debugging purposes

```typescript
debugging: boolean = false;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```typescript
const $ = new Bunbun();
$.logger.debugging = true;
// Now bunbun will throw more logs
$.logger.debugging = false;
// Now bunbun will don't throw so much data
```
</details>

<!------------------------------------------------------------------------------------------------------------->

<h3 id="class-logger--silent"><code>logger.silent</code></h3>

Determines if logger should be silent

```typescript
silent: boolean = false;
```

<details>
  <summary>📚 Click to expand the sample code</summary>

```typescript

CHANGELOG.md
# 3.0.0
    - Reorganizing code
      - Cleaner interface
      - More control over tasks
      - Less same-doing functions
        - `try-` functions are replaced with single catching function
      - Less non-tasks features
        - All non-tasks features are namespaced
    - Tasks debouncing as default behavior

# 2.0.1
    - Splitting code into smaller parts
    - Creating `CHANGELOG` for feature

# 2.0.0
    - Recreating files reading feature
    - Adding temporary directory feature

# 1.x
    - First ideas

main.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = __importDefault(require("util"));
const chalk_1 = __importDefault(require("chalk"));
const chokidar_1 = __importDefault(require("chokidar"));
const path_1 = __importDefault(require("path"));
const koa_1 = __importDefault(require("koa"));
const ws_1 = __importDefault(require("ws"));
const mime_types_1 = require("mime-types");
const hasha_1 = __importDefault(require("hasha"));
const fast_glob_1 = __importDefault(require("fast-glob"));
const cpy_1 = __importDefault(require("cpy"));
const child_process_1 = require("child_process");
const fs_extra_1 = __importDefault(require("fs-extra"));
const tmp_1 = __importDefault(require("tmp"));
const IGNORED_DEFAULT = Symbol('default');
class BunbunHttpServer {
    constructor(_$, _options) {
        this._$ = _$;
        this._http = new koa_1.default();
        this._options = {
            directory: './build',
            fallback: './index.html',
            reload: true,
            port: 8080,
            reloadPort: 8181,
        };
        this._options = Object.assign({}, this._options, _options);
        const dir = path_1.default.resolve(process.cwd(), this._options.directory);
        const fallback = path_1.default.resolve(dir, this._options.fallback);
        this._http.use((ctx, next) => __awaiter(this, void 0, void 0, function* () {
            if (ctx.path === '/__bunbun-reload.js') {
                ctx.body = yield fs_extra_1.default.readFile(path_1.default.join(__dirname, 'inject-reload.js'), 'utf8');
                ctx.body = ctx.body.replace('__PORT__', this._options.reloadPort);
                return;
            }
            yield next();
        }));
        this._http.use((ctx) => __awaiter(this, void 0, void 0, function* () {
            let file = path_1.default.resolve(dir, path_1.default.join('.', ctx.path));
            if ((yield _$.exists(file)) !== 'file') {
                file = path_1.default.resolve(dir, fallback);
            }
            if ((yield _$.exists(file)) !== 'file') {
                return ctx.throw(404, `cannot find file: ${ctx.path}`);
            }
            const isHtml = ['.htm', '.html'].includes(path_1.default.extname(file).toLowerCase());
            let body = '';
            if (isHtml && this._options.reload) {
                body = yield fs_extra_1.default.readFile(file, 'utf8');
                body = body.concat('<script src="/__bunbun-reload.js"></script>');
            }
            else {
                body = yield fs_extra_1.default.readFile(file);
            }
            ctx.body = body;
            ctx.type = mime_types_1.lookup(file) || 'plain/text';
        }));
        if (this._options.reload) {
            this._ws = new ws_1.default.Server({
                port: this._options.reloadPort,
            });
        }
        this._http.listen(this._options.port);
        _$.log('>> web server listening on port %s', this._options.port);
    }
    reload() {
        const { _$, _options, _ws } = this;
        if (!_options.reload) {
            _$.error('>> ! tried to reload server that isn\'t prepared for reload');
            _$.error('>> ! set-up %s with %s in options', '.serve()', '"reload": true');
            return;
        }
        if (!_ws) {
            return;
        }
        for (const c of _ws.clients) {
            c.send('reload');
        }
    }
}
class Bunbun {
    constructor() {
        this._debug = false;
        this._makeTry = (promise, _default, _defaultTrue = IGNORED_DEFAULT) => {
            return new Promise((res, rej) => {
                promise.then(value => {
                    res(_defaultTrue === IGNORED_DEFAULT ? value : _defaultTrue);
                }).catch(err => {
                    if (this._debug) {
                        console.error(err);
                    }
                    res(_default);
                });
            });
        };
        this._tasks = new Map();
        this._currentTempStack = 0;
        this._tempDirs = new Set();
    }
    debug(val) {
        this._debug = val;
    }
    hashFile(file, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            const r = yield hasha_1.default.fromFile(file, opts);
            return r;
        });
    }
    hash(text, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            const r = yield hasha_1.default.async(text, opts);
            return r;
        });
    }
    wait(ms) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise(res => {
                setTimeout(() => {
                    res();
                }, ms);
            });
        });
    }
    tempDir(cleanOnFinish = true) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                tmp_1.default.dir((err, path, cleanup) => {
                    if (err) {
                        rej(err);
                        return;
                    }
                    this._tempDirs.add({
                        path,
                        stack: this._currentTempStack,
                        clean: cleanOnFinish,
                        cleanup,
                    });
                    res(path);
                });
            });
        });
    }
    tempDirClean(path) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise(res => {
                for (const e of this._tempDirs.values()) {
                    if (e.path === path) {
                        e.cleanup();
                        this._tempDirs.delete(e);
                    }
                }
                res();
            });
        });
    }
    _tempStackDown() {
        this._currentTempStack -= 1;
        for (const e of this._tempDirs.values()) {
            if (e.stack > this._currentTempStack && e.clean) {
                e.cleanup();
                this._tempDirs.delete(e);
            }
        }
    }
    _tempStackUp() {
        this._currentTempStack += 1;
    }
    task(name, tasksOrFn) {
        this._tasks.set(name, tasksOrFn);
    }
    remove(path) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fs_extra_1.default.remove(path);
        });
    }
    tryRemove(path) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._makeTry(this.remove(path), true, false);
        });
    }
    read(file) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                fs_extra_1.default.readFile(file, {
                    encoding: 'utf8',
                }, (err, data) => {
                    err ? rej(err) : res(String(data));
                });
            });
        });
    }
    tryRead(file) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._makeTry(this.read(file), '');
        });
    }
    write(file, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                fs_extra_1.default.writeFile(file, data, {
                    encoding: 'utf8',
                }, err => {
                    err ? rej(err) : res();
                });
            });
        });
    }
    tryWrite(file, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._makeTry(this.write(file, data), true, false);
        });
    }
    readRaw(file) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                fs_extra_1.default.readFile(file, (err, data) => {
                    err ? rej(err) : res(data);
                });
            });
        });
    }
    tryReadRaw(file) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._makeTry(this.readRaw(file), Buffer.from(''));
        });
    }
    writeRaw(file, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                fs_extra_1.default.writeFile(file, data, err => {
                    err ? rej(err) : res();
                });
            });
        });
    }
    tryWriteRaw(file, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._makeTry(this.writeRaw(file, data), true, false);
        });
    }
    copy(source, target, overwrite = true) {
        return __awaiter(this, void 0, void 0, function* () {
            yield fs_extra_1.default.copy(source, target, { overwrite });
        });
    }
    tryCopy(source, target, overwrite = true) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._makeTry(this.copy(source, target, overwrite), true, false);
        });
    }
    globCopy(source, target, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            yield cpy_1.default(source, target, opts);
        });
    }
    tryGlobCopy(source, target, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            {
                return yield this._makeTry(this.globCopy(source, target, opts), true, false);
            }
        });
    }
    exec(command, opts = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((res, rej) => {
                child_process_1.exec(command, {
                    encoding: 'buffer',
                    cwd: opts.cwd || process.cwd(),
                    env: Object.assign({}, process.env, opts.env || {}),
                    timeout: opts.timeout,
                    maxBuffer: opts.maxBuffer,
                }, (err, stdout, stderr) => {
                    if (err) {
                        rej(err);
                        return;
                    }
                    res({ stdout, stderr });
                });
            });
        });
    }
    tryExec(command, opts = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._makeTry(this.exec(command, opts), {
                stdout: Buffer.from(''),
                stderr: Buffer.from(''),
            });
        });
    }
    glob(target, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield fast_glob_1.default(target, opts);
            return files;
        });
    }
    watch(target, fn, opts) {
        chokidar_1.default.watch(target, Object.assign({}, {
            ignoreInitial: true,
        }, opts)).on('all', () => {
            fn();
        });
        fn();
    }
    debounce(fn) {
        let promise;
        let running = false;
        let repeat = false;
        const cb = () => {
            return Promise.resolve().then(() => fn()).finally(() => {
                if (repeat) {
                    repeat = false;
                    return cb();
                }
                else {
                    running = false;
                }
            });
        };
        return () => {
            if (running) {
                repeat = true;
                return promise;
            }
            running = true;
            promise = Promise.resolve().then(() => cb());
            return promise;
        };
    }
    exists(path) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const s = yield fs_extra_1.default.stat(path);
                return s.isDirectory() ? 'dir' : 'file';
            }
            catch (e) {
                return false;
            }
        });
    }
    serve(directory, port, options = {}) {
        return new BunbunHttpServer(this, Object.assign({}, options, {
            directory,
            port,
        }));
    }
    run(name) {
        const x = this._tasks.get(name);
        if (x === undefined) {
            this.error('Cannot find task "%s"', name);
            return;
        }
        this._tempStackUp();
        this.log('>> %s start', name);
        let time = Date.now();
        return (Array.isArray(x)
            ? Promise.all(x.map(y => this.run(y)))
            : Promise.resolve().then(() => x())).catch(msg => {
            this.error('>> %s has crashed', name);
            if (msg) {
                console.log(msg);
            }
        }).finally(() => {
            this.log('>> %s done in time %ss', name, (Date.now() - time) / 1000);
            this._tempStackDown();
        });
    }
    log(text, ...params) {
        this._log(text, chalk_1.default.blue, ...params);
    }
    error(text, ...params) {
        this._log(text, chalk_1.default.red, ...params);
    }
    _log(text, color, ...params) {
        console.log(util_1.default.format(text, ...params.map(x => {
            return typeof x === 'string' || typeof x === 'number'
                ? color(x)
                : x;
        })));
    }
}
exports.default = Bunbun;
module.exports = Bunbun;
//# sourceMappingURL=main.js.map